### PR TITLE
AdbDevice handle abstraction

### PR DESCRIPTION
**Pull request by Halastra**
`TcpHandle` is now a subclass of `HandleBase`
`AdbDevice()` now takes a `HandleBase` instance
Original functionality provided by `AdbDevice.from_tcp()`

Refactored `TcpHandle` related `"serial"` references to `"socket_address"`

### CODE REVIEW

**Comment by Halastra** [comment]
I don't know if Travis CI should run automatically.
I ran all of the existing tests locally, for what it's worth.

**Comment by JeffLIrion** [comment]
That's odd that the CI tests didn't get triggered... I'll close this and re-open it to see if they run. 

I've got some modifications to my branch that I haven't pushed yet. 

On the one hand, I see your point about keeping `AdbDevice.__init__` simple in terms of its parameters. On the other hand, I like the idea of a one-and-done constructor. An alternative approach could be to create a subclass `AdbDeviceTcp` of `AdbDevice`, and similarly there could be an `AdbDeviceUsb` subclass once that functionality is implemented.

**Comment by Halastra** [comment]
I see where you're coming from. The apparent elegance of such a constructor is tempting.
But I don't see how your suggested subclass approach would simplify things.
Assuming this project's aim is to support both TCP and USB (like `python-adb`), you're still left with the question of how to choose the right one (e.g. `AdbDeviceTcp` or `AdbDeviceUsb`), and that would have to be resolved by the constructor.

**Bottom line:** This PR is still the way I would do it, but at the end of the day it's your call. 
With great power, comes great responsibility.

Either way, I'll probably keep working on USB support, but I need to know whether to integrate it with this PR in mind or not.

File: adb_shell/tcp_handle.py [hunk 1]
```diff
@@ -55,26 +56,35 @@ class TcpHandle(object):
...
-        self.serial = '{}:{}'.format(self.host, self.port)
+    def __init__(self, host, port=DEFAULT_PORT, default_timeout_s=None):
+        self.host = host
+        self.port = port
 
         self._default_timeout_s = default_timeout_s
 
         self._connection = None
 
+    @classmethod
```
Comment (JeffLIrion) [inline adb_shell/tcp_handle.py]: I think we can get rid of this method and use

```python
def __init(self, host, port=DEFAULT_PORT, default_timeout_s=None):
```

instead of

```python
def __init__(self, serial, default_timeout_s=None):
```

Then instead of checking for a ":" in `serial` and splitting it, we can simply require the user to provide the host and port separately.

**Comment by JeffLIrion** [comment]
I merged in my pull request, so I'm closing this one.  

I am open to the changes that you want to make, but let's make one change at a time.  My pull request added a `handle` parameter to `AdbDevice.__init__`.  I also moved the handle code into a subpackage, designating a clear place for the USB handle code to go.  

The next step is to further generalize `AdbDevice`, which will entail removing the `serial` and `default_timeout_s` parameters and implementing a `from_tcp` classmethod.  Also, based on this pull request, it looks like some of the terminology in `tcp_handle.py` should be revised.  If you submit a new pull request that implements those things, I'll approve it.

File: adb_shell/tcp_handle.py (hunk 1 above)
Comment (Halastra) [reply adb_shell/tcp_handle.py]: Yes, the `from_socket_address` classmethod is completely optional, for convenience (and some compatibility with the original API).
